### ISSUE TITLE
[BUG] Using a combination of `informalfigure` and `variablelist` inside `procedure` breaks the formatting and numbering of the procedure

### ISSUE THREAD

**Issue by r0ckarong** [openSUSE/daps#478]
### Problem description

I'm using a combination of `informalfigure` for screenshots and `variablelist` to explain contents of the screenshots within a procedure `step`. If the `informalfigure` element is not preceded by text but followed by a `variablelist`, somehow this breaks the formatting in the output.

In the example, the third step should have contained the image followed by the term definitions.
What it actually does, is begin the numbering and apparent element boundary after the image.
This then appears like the image belongs to the previous step and formatting for the variables is broken.

![varlistbug](https://user-images.githubusercontent.com/4920542/46029160-f7fd6e80-c0f2-11e8-809c-e783c5bab9e6.png)

(Note how the 3. "floats" next to the Term 1 text, this is not even the correct boundary for this element).

If you simply add a new `para` of text before `informalfigure` it arranges the elements correctly.

![varlistbug2](https://user-images.githubusercontent.com/4920542/46029759-950cd700-c0f4-11e8-8fb0-fb779cf2bab2.png)

### Expected behavior
The numbering element should be displayed at the correct location next to the image even if no preceding text is present.

### Steps to reproduce the behavior

1. Initialize a new daps directory
2. Replace DC-daps-example with the contents of https://paste.opensuse.org/43780991
3. `daps -d DC-daps-example html`
-> Wrong output
4. Uncomment the `para` at line 29
5. Build again
-> Correct formatting

The diff between the resulting html files looks like this:
```
 diff wrong.html right.html
9c9,12
<     </p></li><li class="step"><div class="informalfigure"><div class="mediaobject"><table border="0" summary="manufactured viewport for HTML img" style="cellpadding: 0; cellspacing: 0;" width="30%"><tr><td><img src="images/example1.png" width="100%" /></td></tr></table></div></div><div class="variablelist"><dl class="variablelist"><dt><span class="term">Term 1</span></dt><dd><p>
---
>     </p></li><li class="step"><p>
>      If you add a paragraph of text before the image, the numbering is correct
>      and the broken style does not appear.
>     </p><div class="informalfigure"><div class="mediaobject"><table border="0" summary="manufactured viewport for HTML img" style="cellpadding: 0; cellspacing: 0;" width="30%"><tr><td><img src="images/example1.png" width="100%" /></td></tr></table></div></div><div class="variablelist"><dl class="variablelist"><dt><span class="term">Term 1</span></dt><dd><p>
```

**Comment by tomschr** [comment]
@r0ckarong: Thanks Markus for this detailed bug report! :+1: Much appreciated! As a workaround, you could probably use an empty `<para/>` element for the time being.

Although I agree, this is unfortunate and a layout bug. However, we didn't anticipate this specific use case I suppose. IMHO, this situation goes beyond a simple layout bug and raises the following, related question to me:

> Should we allow/disallow such use case?
> Should we change Geekodoc to have at least one `para` as first child?

In other words, it's more a *styleguide issue* than a *technical issue*.

As a procedure contain steps and each step contains an action, therefor IMHO each step should contain at least one paragraph *before* any image. This sentence contains the action ("Click ABC") and the image presents the result of it.

Showing just the image as a step, wouldn't that be confusing? What should the user do? Is this the result or is this an action to do anything? Do I need to click? Where...?
IMHO, this gives too much room for interpretation and I think we should avoid such situations. With Geekodoc validation we can direct our writers to do the "right structure".

I think, @taroth21 and @sknorr could give some insights. Any ideas?

---

Currently, Geekodoc allows inside the [step element](https://github.com/openSUSE/geekodoc/blob/develop/geekodoc/rng/geekodoc5.rnc#L119-L150) a wide range of other elements. We could restrict that to the following content model:

```rnc
db.step = db.para.blocks, (...)*
```

Which means: inside a `step` element, start with one `para` element, followed by other elements. These other elements can be lists, tables, figures, or even paragraphs and can occur zero or more times.

However, as a downside, that could *possibly* create validation issues. The question is: do we have steps which *do not start* with a `para`?

**Comment by tomschr** [comment]
@r0ckarong: during lunch, we've discussed this bug. :wink: @taroth21 & @sknorr and I agreed, that a step has to start with at least one paragraph and not with an image. An image alone would be too misleading as I've described it in my last post.

I'd recommend to adjust your text accordingly to make sure further versions of Geekodoc do not invalidate your documentation.

For this reason, I've opened two bugs, one for Geekodoc for validation, and one for the styleguide. Both issues tackle this scenario from different sides.

Setting it to "invalid". In case I've missed something, just reopen it. Thanks anyway for helping us to make us aware of this issue. :+1:
